csharp/src/Apache.Arrow/Arrays/ArrayData.cs

```diff
@@ -28,12 +27,25 @@ public sealed class ArrayData : IDisposable
 
         public readonly IArrowType DataType;
         public readonly int Length;
-        public readonly int NullCount;
+        private int _nullCount;
         public readonly int Offset;
         public readonly ArrowBuffer[] Buffers;
         public readonly ArrayData[] Children;
         public readonly ArrayData Dictionary; // Only used for dictionary type
 
+        public int NullCount
```

Comment:
Is this ABI breakage acceptable? I expect most users would access the null count from the Array rather than ArrayData so it might not affect too many users. But as an alternative I could add a new property and mark this one as obsolete, or maybe just always compute the null count in the constructor rather than doing it on demand?

Reply:
Yeah, this is a tough one. The problem with ABI breakage is that it makes diamond-shaped dependency graphs unfixable without modifying some other dependency. Computing eagerly is a potential performance regression. Marking the original field as obsolete and defining a new property feels like it's probably the best option. I suspect most uses will go through IArrowArray.NullCount anyway, so the impact on existing source is likely to be small. The only painful part is probably coming up with a new name... .

Reply:
Yeah OK, maybe a `GetNullCount` method instead would make sense, partly because I can't think of a better property name, but then it's also obvious it might do more work than a plain member access.

I guess we also don't necessarily need to deprecate the existing member because there might be cases where users only want the null count if it can be accessed quickly and don't want to pay the cost of computing it, so we could just document that it may return a negative number. And we could remove the `readonly` modifier to allow updating it if `GetNullCount` is called, which isn't a breaking change according to [the rules](https://learn.microsoft.com/en-us/dotnet/core/compatibility/library-change-rules#members).

Reply:
I've reverted `NullCount` back to a normal member but made it non-readonly and added a `GetNullCount` method instead. I've changed most uses of `NullCount` to `GetNullCount()`, except in the C Data interface implementation which explicitly allows -1 values: https://arrow.apache.org/docs/format/CDataInterface.html#c.ArrowArray.null_count

Let me know if you'd prefer to deprecate `NullCount` instead though, or have any other suggestions for how to handle this.